### PR TITLE
Clock: unit and basic fuzz tests

### PR DESCRIPTION
**Pull request by DonChangfoot**
This updates ideal Deterministic Time to be used in unit and fuzz
tests.

A Mock Network has been introduced that simulates a network where
packets may be delayed, dropped or replayed. Network metrics such as
one-way-delay, probability of packet loss or replays have been
parameterized so that differing network performance may be modeled.
This will be used in future fuzz testing work.

A Clock Simulator has been introduced to fuzz test the clock
implementation. It uses the Mock Network to send packets between the
clocks.

### CODE REVIEW

File: src/mock_network.zig [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        };
+        const Data = struct {
+            expiry: u64,
+            callback: fn (
+                packet: Packet,
+                to: u8,
+                from: u8,
+            ) void,
+            packet: Packet,
+            direction: Direction,
```
Comment (ifreund) [inline src/mock_network.zig:51]: Why do we need this state? We should already have a separate queue for each direction of a connection on our paths slice, one at `self.path(a, b)` and the other at `self.path(b, a)`.

File: src/vr/clock.zig [hunk 2]
```diff
@@ -74,328 +75,337 @@ const Epoch = struct {
...
-    window: Epoch,
-
-    /// A static allocation to convert window samples into tuple bounds for Marzullo's algorithm.
-    marzullo_tuples: []Marzullo.Tuple,
-
-    /// A kill switch to revert to unsynchronized realtime.
-    synchronization_disabled: bool,
+pub fn ClockImpl(comptime ClockType: type) type {
+    return struct {
+        const Self = @This();
```
Comment (ifreund) [inline src/vr/clock.zig:80]: Hmm, I've already made essentially this change to make `Clock` generic over `Time` on the `beta2` branch. Not sure what the best way to resolve this conflict would be.

**Comment by ifreund** [comment]
I think we should probably rebase this onto the `beta2` branch and target it there. I can handle that if you like.

**Comment by DonChangfoot** [comment]
> I think we should probably rebase this onto the `beta2` branch and target it there. I can handle that if you like.

@ifreund  That would be great, thanks.

**Comment by jorangreef** [comment]
@DonChangfoot great work on this and @ifreund too. The network simulator is awesome.

@DonChangfoot do you have bandwidth available to update the clock simulation test (after @ifreund retargets to `beta2`)  to measure and print out how much accuracy or minimum error the synchronization was able to achieve? i.e. The final metric of our tracing analysis.

So we know how far off the clocks are, and then we're making the network a "sea of simulation" (in the words of Tron) and then finally we show our min clock error in `realtime_synchronized`, so that the test can not just test for correctness but also show us how much resolution we're able to achieve (e.g. whether we can correct clocks and sync down to the microsecond level or only to the millisecond level)?

**Comment by ifreund** [comment]
Finshed the rebase onto `beta2`.

**Comment by DonChangfoot** [comment]
thanks @ifreund 

@jorangreef yep, I'll add those tests

**Review by ifreund — APPROVED** [review]
Looks good to me, Thanks! Merging this into `beta2` so we can use the network simulator for VR fuzz testing.

**Pull request merged**